File: app/src/main/java/com/liadpaz/amp/SettingsActivity.java
package com.liadpaz.amp;

import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;
import androidx.preference.SwitchPreference;

import com.liadpaz.amp.databinding.SettingsActivityBinding;
import com.liadpaz.amp.utils.Constants;
import com.liadpaz.amp.utils.LocalFiles;
import com.liadpaz.amp.utils.Utilities;

public class SettingsActivity extends AppCompatActivity {
    private static final int REQUEST_PICK_FOLDER = 44;

    @SuppressWarnings("ConstantConditions")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        com.liadpaz.amp.databinding.SettingsActivityBinding binding;
        setContentView((binding = SettingsActivityBinding.inflate(getLayoutInflater())).getRoot());
        getSupportFragmentManager().beginTransaction().replace(R.id.settings, new SettingsFragment()).commit();
        setSupportActionBar(binding.toolbarSettings);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
    }

    public static class SettingsFragment extends PreferenceFragmentCompat {
        private Preference pathPreference;
        private SwitchPreference notificationPreference;
        private SwitchPreference screenOnPreference;

        private String path;

        @SuppressWarnings("ConstantConditions")
        @Override
        public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
            setPreferencesFromResource(R.xml.root_preferences, rootKey);

            path = LocalFiles.getPath();

            pathPreference = findPreference("path");
            pathPreference.setSummary(TextUtils.isEmpty(path) ? getString(R.string.preference_all_path) : path);
            pathPreference.setOnPreferenceClickListener(preference -> {
                startActivityForResult(new Intent(Intent.ACTION_OPEN_DOCUMENT_TREE).addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION), REQUEST_PICK_FOLDER);
                return true;
            });
            findPreference("default_path").setOnPreferenceClickListener(preference -> {
                LocalFiles.setPath(Constants.DEFAULT_PATH);
                pathPreference.setSummary(Constants.DEFAULT_PATH);
                requireActivity().setResult(path.equals(Constants.DEFAULT_PATH) ? RESULT_CANCELED : RESULT_OK);
                return true;
            });
            notificationPreference = findPreference("notification");
            notificationPreference.setChecked(LocalFiles.getShowCurrent());
            notificationPreference.setOnPreferenceClickListener(preference -> {
                LocalFiles.setShowCurrent(notificationPreference.isChecked());
                return true;
            });
            screenOnPreference = findPreference("screen_on");
            screenOnPreference.setChecked(LocalFiles.getScreenOn());
            screenOnPreference.setOnPreferenceClickListener(preference -> {
                LocalFiles.setScreenOn(screenOnPreference.isChecked());
                return true;
            });
        }

        @SuppressWarnings("ConstantConditions")
        @Override
        public void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
            if (requestCode == REQUEST_PICK_FOLDER && resultCode == RESULT_OK) {
                String path = Utilities.getPathFromUri(data.getData());
                if (!LocalFiles.getPath().equals(path)) {
                    LocalFiles.setPath(path);
                    pathPreference.setSummary(TextUtils.isEmpty(path) ? getString(R.string.preference_all_path) : path);
                    requireActivity().setResult(SettingsFragment.this.path.equals(path) ? RESULT_CANCELED : RESULT_OK);
                }
            }
        }
    }
}

File: app/src/main/java/com/liadpaz/amp/ui/SettingsActivity.kt
package com.liadpaz.amp.ui

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.text.TextUtils
import androidx.appcompat.app.AppCompatActivity
import androidx.preference.CheckBoxPreference
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import androidx.preference.SwitchPreferenceCompat
import com.liadpaz.amp.R
import com.liadpaz.amp.databinding.SettingsActivityBinding
import com.liadpaz.amp.utils.Constants
import com.liadpaz.amp.utils.LocalFiles
import com.liadpaz.amp.utils.Utilities

class SettingsActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        var binding: SettingsActivityBinding
        setContentView(SettingsActivityBinding.inflate(layoutInflater).also { binding = it }.root)
        setSupportActionBar(binding.toolbarSettings)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        supportFragmentManager.beginTransaction().replace(R.id.settings, SettingsFragment()).commit()
    }

    class SettingsFragment : PreferenceFragmentCompat() {

        private lateinit var path: String

        override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
            setPreferencesFromResource(R.xml.root_preferences, rootKey)

            path = LocalFiles.path

            val pathPreference = findPreference<Preference>("path")?.apply {
                summary = if (TextUtils.isEmpty(path)) getString(R.string.preference_all_path) else path
                onPreferenceClickListener = Preference.OnPreferenceClickListener {
                    startActivityForResult(Intent(Intent.ACTION_OPEN_DOCUMENT_TREE).addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION), REQUEST_PICK_FOLDER)
                    true
                }
            }!!
            findPreference<Preference>("default_path")?.apply {
                onPreferenceClickListener = Preference.OnPreferenceClickListener {
                    pathPreference.summary = Constants.DEFAULT_PATH
                    requireActivity().setResult(if (path == Constants.DEFAULT_PATH) Activity.RESULT_CANCELED else Activity.RESULT_OK)
                    true
                }
            }
        }

        override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
            if (requestCode == REQUEST_PICK_FOLDER && resultCode == Activity.RESULT_OK) {
                val path = Utilities.getPathFromUri(data?.data!!)
                if (LocalFiles.path != path) {
                    findPreference<Preference>("path")?.summary = if (TextUtils.isEmpty(path)) getString(R.string.preference_all_path) else path
                    requireActivity().setResult(if (this@SettingsFragment.path == path) Activity.RESULT_CANCELED else Activity.RESULT_OK)
                }
            }
        }
    }

    companion object {
        private const val REQUEST_PICK_FOLDER = 44
    }
}